ENGAGEHF/Managers/UserMetaDataManager/UserMetaDataManager.swift

```diff
@@ -0,0 +1,167 @@
+                    guard let self else {
+                        return
+                    }
+
+                    switch event {
+                    case let .associatedAccount(details):
+                        updateSnapshotListener(for: details)
+                    case .disassociatingAccount:
+                        updateSnapshotListener(for: nil)
+                    default:
+                        break
+                    }
+                }
+            }
+        }
+        
+        if let account {
+            updateSnapshotListener(for: account.details)
+        }
+    }
+    
+    
+    /// Call on change of notification settings in Account Sheet.
+    /// Updates the user's notification settings in Firestore to the current values stored in the manager.
+    func pushUpdatedNotificationSettings() async {
+        self.logger.debug("Updating notification settings.")
+        
+        guard let details = account?.details else {
+            return
+        }
+        
+        print(self.notificationSettings.codingRepresentation)
+        
+        let userDocRef = Firestore.userDocumentReference(for: details.accountId)
+        
+        do {
+            try await userDocRef.updateData(self.notificationSettings.codingRepresentation)
+        } catch {
+            self.logger.error("Failed to update notification settings: \(error)")
+            return
+        }
+    }
+    
+    
+    /// Called on sign-in. Registers a snapshot listener to the user's meta-data document in Firestore.
+    /// Collects information such as notification preferences and organization contact information.
+    private func updateSnapshotListener(for details: AccountDetails?) {
+        self.logger.debug("Initializing user information snapshot listener...")
+        
+        self.snapshotListener?.remove()
+        
+        guard let details else {
+            return
+        }
+        
+        let userDocRef = Firestore.userDocumentReference(for: details.accountId)
+        
+        self.snapshotListener = userDocRef
+            .addSnapshotListener { docSnapshot, error in
+                if let error {
+                    self.logger.error("Failed to initialize user information snapshot: \(error)")
+                    return
+                }
+                
+                self.logger.debug("Fetching user metadata.")
+                
+                guard let userDoc = docSnapshot else {
+                    self.logger.error("No document found in user document snapshot.")
+                    return
+                }
+                
+                // Only fetch organization information from Firestore if we haven't already fetched it
+                // for the current user.
+                if self.organization == nil {
+                    Task { @MainActor in
+                        await self.getOrganizationInfo(from: userDoc)
+                    }
+                }
```

Comment:
What is a user logs out and logs back in with a different organization; wouldn't we have to pull it again here? Might be mistaken and it might also be smarter to reset the organization on user logout in the Spezi Account event handler above?

Reply:
The motivation for this is the fact that because we register a snapshot listener to the user's document in `Users`, every time we change the notification settings from the client we pull the organization data all over again. The idea here was to prevent these unnecessary Firestore reads by only pulling organization information when the user first signs in (assuming the organization won't change frequently enough to warrant listening to these fields).

It could be that the snapshot listener is unnecessary. Initially, I registered a snapshot listener in case the server changes a user's notification settings, and to make sure that the client and the server have the same ground truth. If we don't have a snapshot listener, then we would only fetch the settings and organization info from the server on sign-in. Updating the notification settings on the client would write the current State of the Notification Settings on the client to Firestore. What do you think?

Reply:
In the end, I think the snapshot listener should stay -- if then user has the app on multiple devices, this mechanism will sync changes between them, and it makes sure that the source of truth for notification settings comes from the server (if we blindly write to the server when the client state changes, but the write fails, then if we don't have a snapshot listener then the server and the client will have conflicting settings). I'll go ahead and set the organization to nil on sign out!